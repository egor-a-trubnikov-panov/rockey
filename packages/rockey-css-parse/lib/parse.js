import cssToJSON from './cssToJSON';
import createMixin from './mixins/createMixin';
import interpolateWithMixins from './mixins/interpolateWithMixins';
import * as classnames from './utils/classnames';

const isMedia = key => key.indexOf('@media') === 0;
const isKeyFrames = key => key.indexOf('@keyframes') === 0;
const isNot = key => key.indexOf(':not') === 0;
const NOT_REGEX = /\(([^\)]+)\)(.*)/;

// @replace-start
const isModificatorStartSymbol = symbol => {
  return symbol === '@' || symbol === ':' || symbol === '[';
};
// @replace-end

// @replace-start
const isModificatorEndSymbol = symbol => {
  return symbol === '{' || symbol === ';';
};
// @replace-end

const isStartsWithModificator = raw => {
  return (
    0 === raw.indexOf('@media') ||
    0 === raw.indexOf('[') ||
    0 === raw.indexOf('@keyframes') ||
    0 === raw.indexOf('::placeholder') ||
    0 === raw.indexOf('::after') ||
    0 === raw.indexOf('::before') ||
    0 === raw.indexOf('::first-letter') ||
    0 === raw.indexOf('::first-line') ||
    0 === raw.indexOf(':active') ||
    0 === raw.indexOf(':checked') ||
    0 === raw.indexOf(':disabled') ||
    0 === raw.indexOf(':empty') ||
    0 === raw.indexOf(':enabled') ||
    0 === raw.indexOf(':first-child') ||
    0 === raw.indexOf(':first-of-type') ||
    0 === raw.indexOf(':focus') ||
    0 === raw.indexOf(':hover') ||
    0 === raw.indexOf(':in-range') ||
    0 === raw.indexOf(':invalid') ||
    0 === raw.indexOf(':lang') ||
    0 === raw.indexOf(':last-child') ||
    0 === raw.indexOf(':last-of-type') ||
    0 === raw.indexOf(':link') ||
    0 === raw.indexOf(':not') ||
    0 === raw.indexOf(':nth-child') ||
    0 === raw.indexOf(':nth-last-child') ||
    0 === raw.indexOf(':nth-last-of-type') ||
    0 === raw.indexOf(':nth-of-type') ||
    0 === raw.indexOf(':only-of-type') ||
    0 === raw.indexOf(':only-child') ||
    0 === raw.indexOf(':optional') ||
    0 === raw.indexOf(':out-of-range') ||
    0 === raw.indexOf(':read-only') ||
    0 === raw.indexOf(':read-write') ||
    0 === raw.indexOf(':required') ||
    0 === raw.indexOf(':root') ||
    0 === raw.indexOf(':target') ||
    0 === raw.indexOf(':valid') ||
    0 === raw.indexOf(':visited')
  );
};

const shouldGenerateSelectors = parent => parent.type !== 'keyframes';

export default function createParser(config = {}) {
  const plugins = config.plugins || [];

  const actions = {
    // getClassName: classnames.getClassName(config.getClassName),
    getMixinClassName: classnames.getMixinClassName(config.getMixinClassName),
    // getSelector: classnames.getSelector(config.getClassName),
    getAnimationName: config.getAnimationName
      ? config.getAnimationName
      : name => name,
  };

  const getComponentName = classnames.getComponentName;

  return function parse(strings, ...values) {
    const context = {
      hasAnimations: false,
      animations: {},
      classnames: {
        components: {},
        selector: {},
      },
    };

    const classname = classnames.getClassName(config.getClassName);
    const getClassName = component => {
      if (context.classnames.components[component]) {
        return context.classnames.components[component];
      }
      const className = classname(component);
      context.classnames.components[component] = className;

      return className;
    };

    const selector = classnames.getSelector(getClassName);
    const getSelector = component => {
      if (context.classnames.selector[component]) {
        return context.classnames.selector[component];
      }
      const className = selector(component);
      context.classnames.selector[component] = className;

      return className;
    };

    let precss = [];

    let raw = null;
    let mixinsFunctions = {};

    if (values.length === 0) {
      raw = Array.isArray(strings) ? strings[0] : strings;
    } else {
      const interpolated = interpolateWithMixins(strings, ...values);
      raw = interpolated.raw;
      mixinsFunctions = interpolated.mixinsFunctions;
    }

    const inline = raw.replace(/\r|\n/g, '').replace(/\s+/g, ' ');

    const root = toPreCSS(inline, {
      type: 'root',
      selector: [],
      media: null,
      root: [],
    });

    return {
      classList: root.reduce((components, c) => {
        components[getComponentName(c)] = getClassName(c);
        return components;
      }, {}),
      precss,
    };

    function toPreCSS(raw, parent = {}) {
      let openedBrackets = 0;
      let openedModificatorBrackets = 0;
      let styles = '';

      let rootComponents = [];
      let components = [];

      let possibleModificator = 0;
      let modificator = '';

      let current = '';
      let currentBackup = '';

      // @replace-start
      const restore = () => {
        current = currentBackup + current;
        currentBackup = '';
      };
      // @replace-end

      // @replace-start
      const clearAndRestore = () => {
        current = currentBackup;
        currentBackup = '';
      };
      // @replace-end

      // @replace-start
      const backup = () => {
        currentBackup = current;
        current = '';
      };
      // @replace-end

      // @replace-start
      const possibleModificatorStart = () => {
        possibleModificator = true;
      };
      // @replace-end

      // @replace-start
      const correctModificatorPosition = () => {
        return (s => !s.length || s.slice(-1) === ';' || s.slice(-1) === '{')(
          current.trim()
        );
      };
      // @replace-end

      // @replace-start
      const isPossibleModificator = () => {
        return possibleModificator;
      };
      // @replace-end

      // @replace-start
      const possibleModificatorEnd = () => {
        possibleModificator = false;
      };
      // @replace-end

      // @replace-start
      const isModificator = () => {
        return isStartsWithModificator(current.trim());
      };
      // @replace-end

      // @replace-start
      const openBracket = () => {
        openedBrackets++;
      };
      // @replace-end

      // @replace-start
      const closeBracket = () => {
        openedBrackets--;
      };
      // @replace-end

      // @replace-start
      const isAllBracketsClosed = () => {
        return openedBrackets === 0;
      };
      // @replace-end

      // @replace-start
      const openModificatorBracket = () => {
        openedModificatorBrackets++;
      };
      // @replace-end

      // @replace-start
      const closeModificatorBracket = () => {
        openedModificatorBrackets--;
      };
      // @replace-end

      // @replace-start
      const isAllModificatorBracketsClosed = () => {
        return openedModificatorBrackets === 0;
      };
      // @replace-end

      // @replace-start
      const startModificator = () => {
        modificator = current;
        current = '';
      };
      // @replace-end

      // @replace-start
      const saveModificator = () => {
        modificator = modificator.trim();

        let selector = [];
        const media = isMedia(modificator);
        const keyframes = isKeyFrames(modificator);

        if (shouldGenerateSelectors(parent)) {
          if (media || keyframes) {
            selector = [...parent.selector];
          } else {
            modificator.split(',').forEach(m => {
              m = m.trim();

              if (isNot(m)) {
                const matches = m.match(NOT_REGEX);
                m = `:not(${getSelector(matches[1])})${matches[2]}`;
              }

              if (parent.selector.length) {
                parent.selector.forEach(s => {
                  selector.push(s + m);
                });
              } else {
                selector.push(m);
              }
            });
          }
        }

        if (keyframes) {
          let animationName = modificator.split(' ')[1];

          let uniqAnimationName = actions.getAnimationName(animationName);

          context.hasAnimations = true;
          context.animations[animationName] = uniqAnimationName;

          const backup = precss;
          precss = [];

          toPreCSS(current, {
            type: 'keyframes',
            root: parent.root,
          });

          backup.push({
            media: media ? modificator : parent.media,
            selector: modificator.replace(animationName, uniqAnimationName),
            frames: precss,
          });

          precss = backup;
        } else {
          toPreCSS(current, {
            type: 'modificator',
            media: media ? modificator : parent.media,
            selector,
            root: parent.root,
          });
        }

        modificator = '';
      };
      // @replace-end

      // @replace-start
      const isInsideModificator = () => {
        return !!modificator;
      };
      // @replace-end

      // @replace-start
      const isRootComponentBlockStarts = symbol => {
        return symbol === '{' && isAllBracketsClosed();
      };
      // @replace-end

      // @replace-start
      const startComponent = () => {
        let i = null;
        for (i = current.length; i > 0; i--) {
          const symbol = current[i];
          if (symbol === ';') {
            i++;
            break;
          }
        }

        let currentComponents = [];

        if (i) {
          styles += current.slice(0, i);
          currentComponents = current.slice(i).split(',');
        } else {
          currentComponents = current.split(',');
        }

        currentComponents.forEach(c => {
          components.push(c.trim());
        });

        current = '';
      };
      // @replace-end

      // @replace-start
      const isInsideComponent = () => {
        return !!components.length;
      };
      // @replace-end

      // @replace-start
      const save = symbol => {
        current += symbol;
      };
      // @replace-end

      // @replace-start
      const saveComponent = () => {
        const selector = [];

        if (shouldGenerateSelectors(parent)) {
          components.forEach(c => {
            rootComponents.push(c);

            const className = getSelector(c);

            if (parent.selector.length) {
              parent.selector.forEach(p => {
                selector.push(`${p} ${className}`);
              });
            } else {
              selector.push(className);
            }
          });
        }

        toPreCSS(current, {
          type: 'component',
          media: parent.media,
          root: parent.root.length ? parent.root : components,
          selector: parent.type === 'keyframes' ? components : selector,
        });

        components = [];
        current = '';
      };
      // @replace-end

      // @replace-start
      const updateStylesTail = () => {
        if (current) {
          styles += current;
        }
      };
      // @replace-end

      // ---
      const mixins = [];
      let possibleMixin = false;
      let mixin = '';

      // @replace-start
      const isMixinStartSymbol = symbol => {
        return symbol === '_';
      };
      // @replace-end

      // @replace-start
      const isMixinEndSymbol = symbol => {
        return symbol === ';' || symbol === ' ';
      };
      // @replace-end

      // @replace-start
      const possibleMixinStart = () => {
        possibleMixin = true;
      };
      // @replace-end

      // @replace-start
      const possibleMixinEnd = () => {
        possibleMixin = false;
      };
      // @replace-end

      // @replace-start
      const isPossibleMixin = () => {
        return possibleMixin;
      };
      // @replace-end

      // @replace-start
      const saveMixinString = symbol => {
        mixin += symbol;
      };
      // @replace-end

      // @replace-start
      const clearMixin = () => {
        mixin = '';
      };
      // @replace-end

      // @replace-start
      const removeMixinFromStyles = () => {
        current = current.replace(mixin, '');
      };
      // @replace-end

      // @replace-start
      const saveMixin = () => {
        const mixinFunction = createMixin({
          className: actions.getMixinClassName(mixinsFunctions[mixin]),
          selector: parent.selector,
          root: parent.root,
          func: mixinsFunctions[mixin],
          parse,
        });

        mixins.push(mixinFunction);
      };
      // @replace-end
      let size = raw.length;
      for (let i = 0; i < size; i++) {
        const symbol = raw[i];

        if (!isInsideComponent() && !isInsideModificator()) {
          if (isMixinStartSymbol(symbol)) {
            possibleMixinStart();
          }

          if (isPossibleMixin()) {
            if (isMixinEndSymbol(symbol)) {
              possibleMixinEnd();
              saveMixin();
              removeMixinFromStyles();
              clearMixin();
              continue;
            } else {
              saveMixinString(symbol);
            }
          }

          if (
            !isPossibleModificator() &&
            isModificatorStartSymbol(symbol) &&
            correctModificatorPosition()
          ) {
            possibleModificatorStart();
            backup();
          }

          if (isPossibleModificator() && isModificatorEndSymbol(symbol)) {
            if (isModificator()) {
              startModificator();
            } else {
              restore();
            }

            possibleModificatorEnd();
          }
        }

        if (
          !isPossibleModificator() &&
          !isInsideModificator() &&
          isRootComponentBlockStarts(symbol)
        ) {
          openBracket();
          startComponent();
          // continue;
        } else if (symbol === '{') {
          if (isInsideComponent()) {
            openBracket();

            if (openedBrackets !== 1) {
              save(symbol);
            }
          } else if (isInsideModificator()) {
            openModificatorBracket();

            if (openedModificatorBrackets !== 1) {
              save(symbol);
            }
          }
        } else if (symbol === '}') {
          if (isInsideComponent()) {
            closeBracket();
          } else if (isInsideModificator()) {
            closeModificatorBracket();
          }

          if (isInsideComponent()) {
            if (isAllBracketsClosed()) {
              saveComponent();
            } else {
              save(symbol);
            }
          } else if (isInsideModificator()) {
            if (isAllModificatorBracketsClosed()) {
              saveModificator();
              clearAndRestore();
            } else {
              save(symbol);
            }
          }
        } else {
          save(symbol);
        }
      }

      restore();
      updateStylesTail();

      styles = styles.trim();

      if (styles || mixins.length) {
        precss.push({
          selector: parent.selector,
          media: parent.media,
          root: parent.root,
          styles: styles ? cssToJSON(styles, context, plugins) : {},
          mixins,
        });
      }

      return rootComponents;
    }
  };
}
