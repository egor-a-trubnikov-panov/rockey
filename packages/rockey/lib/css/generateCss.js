import mergeCss from './utils/mergeCss';

import {
  getClassName,
  getRootClassNameMap,
  generateAnimationName,
} from './getClassName';

const getSelector = (displayName, parent, mixin) => {
  if (
    displayName[0] !== '[' &&
    displayName[0] !== ':' &&
    displayName[0] !== '.' &&
    displayName[0] !== '#' &&
    displayName[0] === displayName[0].toUpperCase()
  ) {
    return `${mixin || ''}${getClassName(displayName, parent)}`;
  } else {
    return `${mixin || ''}${parent ? `${parent} ${displayName}` : `${displayName}`}`;
  }
};

const isMedia = key => key.indexOf('@media') === 0;
const isKeyFrames = key => key.indexOf('@keyframes') === 0;
const isNot = key => key.indexOf(':not') === 0;

// TODO: remove code duplic.
function process(styles, context) {
  let processed = styles;

  if (styles['animation-name']) {
    processed = { ...styles };

    let animationName = styles['animation-name'];
    if (!context.animations[animationName]) {
      if (animationName !== 'none' && animationName.indexOf('none ') !== 0) {
        console.warn(
          `(generate error) "${animationName}" animation was requested but had not been defined`
        );

        delete processed['animation-name'];
      }
    } else {
      processed['animation-name'] = context.animations[animationName];
    }
  } else if (styles['animation']) {
    processed = { ...styles };

    let animationName = styles['animation'].split(' ')[0];

    if (!context.animations[animationName]) {
      if (animationName !== 'none') {
        console.warn(
          `(generate error) "${animationName}" animation was requested but had not been defined`
        );

        delete processed.animation;
      }
    } else {
      processed['animation'] = styles['animation'].replace(
        animationName,
        context.animations[animationName]
      );
    }
  }

  return processed;
}

function processModificators(
  tree,
  {
    parent,
    context,
  }
) {
  const css = {};

  for (const modificatorKey of Object.keys(tree.modificators)) {
    const modificator = tree.modificators[modificatorKey];

    if (isMedia(modificatorKey)) {
      css[modificatorKey] = generateCss(modificator, {
        parent,
      });

      if (Object.keys(modificator.styles).length) {
        css[modificatorKey][parent] = modificator.styles;
      }
    } else if (isKeyFrames(modificatorKey)) {
      let animationName = modificatorKey.split(' ')[1];

      let uniqAnimationName = generateAnimationName(animationName);
      context.animations[animationName] = uniqAnimationName;

      const keyframes = Object.keys(modificator.components).reduce(
        (steps, step) => {
          steps[step] = modificator.components[step].styles;
          return steps;
        },
        {}
      );

      css[modificatorKey.replace(animationName, uniqAnimationName)] = keyframes;
    } else {
      let updatedModificatorKey = modificatorKey;

      if (isNot(modificatorKey)) {
        const matches = modificatorKey.match(/\(([^\)]+)\)(.*)/);
        updatedModificatorKey = `:not(${getSelector(matches[1])})${matches[2]}`;
      }

      // merge?
      if (Object.keys(modificator.styles).length) {
        css[`${parent}${updatedModificatorKey}`] = process(
          modificator.styles,
          context
        );
      }

      Object.assign(
        css,
        generateCss(modificator, {
          parent: `${parent}${updatedModificatorKey}`,
        })
      );
    }
  }

  return css;
}

function generateCss(
  tree,
  {
    parent,
    mixin,
    context,
  }
) {
  const css = {};

  if (parent && Object.keys(tree.modificators)) {
    // render parent modificators
    mergeCss(
      css,
      processModificators(tree, {
        context,
        parent,
      })
    );
  }

  for (const className of Object.keys(tree.components)) {
    const component = tree.components[className];
    const selector = getSelector(className, parent, mixin);

    mergeCss(
      css,
      generateCss(component, {
        parent: selector,
        context,
      })
    );

    if (Object.keys(component.styles).length) {
      css[selector] = process(component.styles, context);
    }
  }

  return css;
}

export default (
  tree,
  {
    isRoot = true,
    parent = null,
    mixin = null,
    context = {
      animations: {},
    },
  } = {}
) => {
  if (Object.keys(tree.styles).length) {
    throw new Error('(generate css) root tree should not contain styles');
  }

  const rootCss = processModificators(tree, {
    context,
    parent,
  });

  const classNameMap = getRootClassNameMap(tree, {
    isRoot,
  });

  return {
    css: mergeCss(
      rootCss,
      generateCss(tree, {
        parent,
        mixin,
        context,
      })
    ),
    context,
    classNameMap: classNameMap,
  };
};
