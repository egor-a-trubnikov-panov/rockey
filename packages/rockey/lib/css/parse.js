// const validateRule = (rule, value) => {};
// const autoprefixRule = (rule, value) => {};

const cache = new Map();

const parseCss = styles => {
  const stylesParts = styles
    .trim()
    .split(';')
    .map(s => s ? s.split(':') : null)
    .filter(s => !!s);

  const key = stylesParts.toString();

  if (!key.trim()) {
    return {};
  }

  if (cache.has(key)) {
    return cache.get(key);
  }

  const stylesJson = stylesParts.reduce(
    (json, value) => {
      if (value) {
        const key = value[0].trim();
        if (json[key]) {
          // css fallback values
          let prev = Array.isArray(json[key]) ? json[key] : [json[key]];

          json[value[0].trim()] = prev.concat([
            value.slice(1).join(':').trim(),
          ]);
        } else {
          json[value[0].trim()] = value.slice(1).join(':').trim();
        }

        // validateRule(value[0], json[value[0]]);
        // json[value[0].trim()] = autoprefixRule(value.slice(1).join(':').trim());
      }
      return json;
    },
    {}
  );

  cache.set(key, stylesJson);

  return stylesJson;
};

// @replace-start
const isModificatorStartSymbol = symbol => {
  return symbol === '@' || symbol === ':' || symbol === '[';
};
// @replace-end

// @replace-start
const isModificatorEndSymbol = symbol => {
  return symbol === '{' || symbol === ';';
};
// @replace-end

// TODO: make this func beauty :)
const isStartsWithModificator = raw => {
  return 0 === raw.indexOf('@media') ||
    0 === raw.indexOf('@keyframes') ||
    0 === raw.indexOf('::after') ||
    0 === raw.indexOf('::before') ||
    0 === raw.indexOf('::first-letter') ||
    0 === raw.indexOf('::first-line') ||
    0 === raw.indexOf(':active') ||
    0 === raw.indexOf(':checked') ||
    0 === raw.indexOf(':disabled') ||
    0 === raw.indexOf(':empty') ||
    0 === raw.indexOf(':enabled') ||
    0 === raw.indexOf(':first-child') ||
    0 === raw.indexOf(':first-of-type') ||
    0 === raw.indexOf(':focus') ||
    0 === raw.indexOf(':hover') ||
    0 === raw.indexOf(':in-range') ||
    0 === raw.indexOf(':invalid') ||
    0 === raw.indexOf(':lang') ||
    0 === raw.indexOf(':last-child') ||
    0 === raw.indexOf(':last-of-type') ||
    0 === raw.indexOf(':link') ||
    0 === raw.indexOf(':not') ||
    0 === raw.indexOf(':nth-child') ||
    0 === raw.indexOf(':nth-last-child') ||
    0 === raw.indexOf(':nth-last-of-type') ||
    0 === raw.indexOf(':nth-of-type') ||
    0 === raw.indexOf(':only-of-type') ||
    0 === raw.indexOf(':only-child') ||
    0 === raw.indexOf(':optional') ||
    0 === raw.indexOf(':out-of-range') ||
    0 === raw.indexOf(':read-only') ||
    0 === raw.indexOf(':read-write') ||
    0 === raw.indexOf(':required') ||
    0 === raw.indexOf(':root') ||
    0 === raw.indexOf(':target') ||
    0 === raw.indexOf(':valid') ||
    0 === raw.indexOf(':visited');
};

const parse = (raw, parent) => {
  let openedBrackets = 0;
  let openedModificatorBrackets = 0;
  let styles = '';

  let component = null;
  let combinedComponents = [];
  let components = {};

  let possibleModificator = 0;
  let modificators = {};
  let modificator = '';

  let current = '';
  let currentBackup = '';

  // @replace-start
  const restore = () => {
    current = currentBackup + current;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const clearAndRestore = () => {
    current = currentBackup;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const backup = () => {
    currentBackup = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const possibleModificatorStart = () => {
    possibleModificator = true;
  };
  // @replace-end

  // @replace-start
  const isPossibleModificator = () => {
    return possibleModificator;
  };
  // @replace-end

  // @replace-start
  const possibleModificatorEnd = () => {
    possibleModificator = false;
  };
  // @replace-end

  // @replace-start
  const isModificator = () => {
    return isStartsWithModificator(current.trim());
  };
  // @replace-end

  // @replace-start
  const openBracket = () => {
    openedBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeBracket = () => {
    openedBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllBracketsClosed = () => {
    return openedBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const openModificatorBracket = () => {
    openedModificatorBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeModificatorBracket = () => {
    openedModificatorBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllModificatorBracketsClosed = () => {
    return openedModificatorBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const startModificator = () => {
    modificator = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const saveModificator = () => {
    modificator = modificator.trim();

    // @remove-start
    if (modificators[modificator]) {
      throw new Error(
        `(parse error) "${modificator}" duplicated definiton at one block`
      );
    }
    // @remove-end

    modificators[modificator] = parse(current, {
      parentType: 'modificator',
      name: modificator,
    });

    modificator = '';
  };
  // @replace-end

  // @replace-start
  const isInsideModificator = () => {
    return !!modificator;
  };
  // @replace-end

  // @replace-start
  const isRootComponentBlockStarts = symbol => {
    return symbol === '{' && isAllBracketsClosed();
  };
  // @replace-end

  // @replace-start
  const startComponent = () => {
    let parts = current.trim().split(' ');
    if (parts.length === 1) {
      component = parts[0];
    } else {
      combinedComponents = [];

      let i = 0;
      for (i = parts.length - 1; i >= 0; i--) {
        const part = parts[i];
        const charCode = part.charCodeAt(0);

        /* All chars starts with uppercase or xxx% for animations  */
        if (
          (charCode >= 65 && charCode <= 90) || part[part.length - 1] === '%'
        ) {
          combinedComponents.push(part.replace(',', '').trim());
        } else if (part[0] === '~' || part[0] === '+') {
          let size = combinedComponents.length;
          combinedComponents[size - 1] = part[0] + combinedComponents[size - 1];
        } else {
          break;
        }
      }

      if (combinedComponents.length) {
        component = combinedComponents[0];
        combinedComponents = combinedComponents.slice(1);
        parts = parts.slice(0, i + 1);
      } else {
        component = parts[parts.length - 1];
        parts = parts.slice(0, -1);
      }

      styles += parts.join(' ');
    }

    current = '';
  };
  // @replace-end

  // @replace-start
  const isInsideComponent = () => {
    return !!component;
  };
  // @replace-end

  // @replace-start
  const save = symbol => {
    current += symbol;
  };
  // @replace-end

  // @replace-start
  const saveComponent = () => {
    // @remove-start
    if (components[component]) {
      throw new Error(
        `(parse error) "${component}" duplicated definiton at one block`
      );
    }
    // @remove-end
    components[component] = parse(current, {
      parentType: 'component',
      name: component,
      combinedComponents,
    });

    combinedComponents = [];

    component = null;
    current = '';
  };
  // @replace-end

  // @replace-start
  const updateStylesTail = () => {
    if (current) {
      styles += current;
    }
  };
  // @replace-end

  // ---
  const mixins = [];
  let possibleMixin = false;
  let mixin = '';

  // @replace-start
  const isMixinStartSymbol = symbol => {
    return symbol === '_';
  };
  // @replace-end

  // @replace-start
  const isMixinEndSymbol = symbol => {
    return symbol === ' ';
  };
  // @replace-end

  // @replace-start
  const possibleMixinStart = () => {
    possibleMixin = true;
  };
  // @replace-end

  // @replace-start
  const possibleMixinEnd = () => {
    possibleMixin = false;
  };
  // @replace-end

  // @replace-start
  const isPossibleMixin = () => {
    return possibleMixin;
  };
  // @replace-end

  // @replace-start
  const saveMixinString = symbol => {
    mixin += symbol;
  };
  // @replace-end

  // @replace-start
  const clearMixin = () => {
    mixin = '';
  };
  // @replace-end

  // @replace-start
  const removeMixinFromStyles = () => {
    current = current.replace(mixin.trim(), '');
  };
  // @replace-end

  // @replace-start
  const saveMixin = () => {
    mixins.push(mixin.trim());
  };
  // @replace-end

  // @remove-start
  const isInValidMixinPosition = () => {
    return !parent ||
      (currentBackup.trim().length &&
        !new RegExp(`(;|{)\\s+?${mixin.trim()}`).test(`${current}`.trim()));
  };
  // @remove-end

  // ----

  for (let symbol of raw) {
    if (!isInsideComponent() && !isInsideModificator()) {
      if (isMixinStartSymbol(symbol)) {
        possibleMixinStart();
      }

      if (isPossibleMixin()) {
        saveMixinString(symbol);
        if (isMixinEndSymbol(symbol)) {
          possibleMixinEnd();
          // @remove-start
          if (isInValidMixinPosition()) {
            if (parent) {
              const nearestCode = `${currentBackup}${current}`
                .replace(/\s+/g, ' ')
                // .replace(mixin.trim(), '{{ MIXIN }}' + mixin)
                .trim();

              console.warn(
                'seems wrong mixin position',
                `${parent.name} - ${nearestCode}`
              );
            } else {
              console.warn('mixin should not be at root');
            }
          }
          // @remove-end

          saveMixin();
          removeMixinFromStyles();
          clearMixin();
        }
      }
    }

    if (!isInsideComponent() && !isInsideModificator()) {
      if (!isPossibleModificator() && isModificatorStartSymbol(symbol)) {
        possibleModificatorStart();
        backup();
      }

      if (isPossibleModificator() && isModificatorEndSymbol(symbol)) {
        if (isModificator()) {
          startModificator();
        } else {
          restore();
        }

        possibleModificatorEnd();
      }
    }

    if (
      !isPossibleModificator() &&
      !isInsideModificator() &&
      isRootComponentBlockStarts(symbol)
    ) {
      openBracket();
      startComponent();
      // continue;
    } else if (symbol === '{') {
      if (isInsideComponent()) {
        openBracket();

        if (openedBrackets !== 1) {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        openModificatorBracket();

        if (openedModificatorBrackets !== 1) {
          save(symbol);
        }
      }
    } else if (symbol === '}') {
      if (isInsideComponent()) {
        closeBracket();
      } else if (isInsideModificator()) {
        closeModificatorBracket();
      }

      if (isInsideComponent()) {
        if (isAllBracketsClosed()) {
          saveComponent();
        } else {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        if (isAllModificatorBracketsClosed()) {
          saveModificator();
          clearAndRestore();
        } else {
          save(symbol);
        }
      }
    } else {
      save(symbol);
    }
  }

  restore();
  updateStylesTail();

  return {
    mixins,
    components,
    combinedComponents: (parent && parent.combinedComponents) || [],
    styles: parseCss(styles),
    modificators,
  };
};

export default inline => {
  return parse(inline.replace(/\r|\n/g, '').replace(/\s+/g, ' '));
};
