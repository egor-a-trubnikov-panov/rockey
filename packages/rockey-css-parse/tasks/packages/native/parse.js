import { argv } from 'yargs';
import chalk from 'chalk';

import getNativeCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import createParser from '../../../lib/parse';
import { stringifyRules } from '../../../lib/stringify';

const size = argv.size || 1000;
const css = getNativeCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

const parse = createParser();

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
