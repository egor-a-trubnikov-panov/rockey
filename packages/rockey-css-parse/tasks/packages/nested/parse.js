import { argv } from 'yargs';
import chalk from 'chalk';

import getCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import createParser from '../../../lib/parse';
import { stringifyRules } from '../../../lib/stringify';

const size = argv.size || 10;
const css = getCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

const parse = createParser();

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
