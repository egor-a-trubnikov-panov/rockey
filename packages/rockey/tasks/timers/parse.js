import { argv } from 'yargs';
import chalk from 'chalk';

import getCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import parse from '../../lib/css/parse';
import { stringifyRules } from '../../lib/styleSheets/utils/stringify';

const size = argv.size || 10;
const css = getCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
