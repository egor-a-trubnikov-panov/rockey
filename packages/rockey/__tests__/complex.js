import parse from '../lib/css/parse';
import generateCss from '../lib/css/generateCss';

jest.mock('../lib/utils/hash', () => {
  return () => '{{ hash }}';
});

test('complex', () => {
  const parsed = parse(
    `
    Button {
      :not([disabled]):hover {
        background-color: transparent
      }
    }
  `
  );

  expect(parsed).toMatchSnapshot();
  const generated = generateCss(parsed);

  expect(generated.css).toMatchSnapshot();
});

test('coma in modificators', () => {
  const parsed = parse(
    `
    Layer {
      ::after, ::before {
        color: red;
      }
    }
  `
  );

  expect(parsed).toMatchSnapshot();
  const generated = generateCss(parsed);

  expect(generated.css).toMatchSnapshot();
});

test('complex selectors', () => {
  const parsed = parse(
    `
    Bar {
      font-size: 10px;

      + NextChild {
        color: red'
      }

      ~ NextFirst {
        padding: 20px;
      }

      color: green;

      WarningButton, PrimaryButton, SucceessButton {
        padding: 10px;
      }
    }
  `
  );

  // expect(parsed).toMatchSnapshot();
  const generated = generateCss(parsed);

  expect(generated.css).toMatchSnapshot();
});
