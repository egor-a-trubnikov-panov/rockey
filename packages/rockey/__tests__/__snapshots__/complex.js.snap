// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coma in modificators 1`] = `
Object {
  "combinedComponents": Array [],
  "components": Object {
    "Layer": Object {
      "combinedComponents": Array [],
      "components": Object {},
      "mixins": Array [],
      "modificators": Object {
        "::after, ::before": Object {
          "combinedComponents": Array [],
          "components": Object {},
          "mixins": Array [],
          "modificators": Object {},
          "styles": Object {
            "color": "red",
          },
        },
      },
      "styles": Object {},
    },
  },
  "mixins": Array [],
  "modificators": Object {},
  "styles": Object {},
}
`;

exports[`coma in modificators 2`] = `
Object {
  ".Layer-{{ hash }}::after": Object {
    "color": "red",
  },
  ".Layer-{{ hash }}::before": Object {
    "color": "red",
  },
}
`;

exports[`complex 1`] = `
Object {
  "combinedComponents": Array [],
  "components": Object {
    "Button": Object {
      "combinedComponents": Array [],
      "components": Object {},
      "mixins": Array [],
      "modificators": Object {
        ":not([disabled]):hover": Object {
          "combinedComponents": Array [],
          "components": Object {},
          "mixins": Array [],
          "modificators": Object {},
          "styles": Object {
            "background-color": "transparent",
          },
        },
      },
      "styles": Object {},
    },
  },
  "mixins": Array [],
  "modificators": Object {},
  "styles": Object {},
}
`;

exports[`complex 2`] = `
Object {
  ".Button-{{ hash }}:not([disabled]):hover": Object {
    "background-color": "transparent",
  },
}
`;

exports[`complex selectors 1`] = `
Object {
  ".Bar-{{ hash }}": Object {
    "color": "green",
    "font-size": "10px",
  },
  ".Bar-{{ hash }} +.NextChild-{{ hash }}": Object {
    "color": "red'",
  },
  ".Bar-{{ hash }} .SucceessButton-{{ hash }},.Bar-{{ hash }} .PrimaryButton-{{ hash }},.Bar-{{ hash }} .WarningButton-{{ hash }}": Object {
    "padding": "10px",
  },
  ".Bar-{{ hash }} ~.NextFirst-{{ hash }}": Object {
    "padding": "20px",
  },
}
`;
